### PR TITLE
feat: on chain router

### PR DESCRIPTION
**Pull request by dmoka**
## Description

Extending the route-executor pallet with on-chain route storing functionality.

Usages:
- Using on-chain route in DCA
- Using on-chain route in Route-Executor if no route is specified
- ExchangeAsset should use on-chain route.

## Checklist:

- [x] I have updated the documentation if necessary.
- [x] I have added tests to cover my changes, regression test if fixing an issue.
- [ ] This is a breaking change.

### CODE REVIEW

File: pallets/route-executor/src/lib.rs [hunk 1]
```diff
@@ -465,3 +702,26 @@ macro_rules! handle_execution_error {
 		}
 	}};
 }
+
+impl<T: Config> RouteProvider<T::AssetId> for Pallet<T> {
+	fn get(asset_pair: AssetPair<T::AssetId>) -> Vec<Trade<T::AssetId>> {
+		let onchain_route = Routes::<T>::get(asset_pair.ordered_pair());
+
+		let default_route = vec![Trade {
```
Comment (mrq1911) [inline pallets/route-executor/src/lib.rs:710]: nice, so you are keeping our current fallback alive

Comment (mrq1911) [reply pallets/route-executor/src/lib.rs:710]: that probably means that you should actually test against it if the omnipool route is not already set in router and not just take any route

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: pallets/omnipool/src/router_execution.rs [hunk 2]
```diff
@@ -148,4 +148,18 @@ impl<T: Config> TradeExecution<OriginFor<T>, T::AccountId, T::AssetId, Balance>
...
+	fn get_liquidity_depth(
+		pool_type: PoolType<T::AssetId>,
+		asset_a: T::AssetId,
+		_asset_b: T::AssetId,
+	) -> Result<Balance, ExecutorError<Self::Error>> {
+		if pool_type != PoolType::Omnipool {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let asset_state = Self::load_asset_state(asset_a).map_err(ExecutorError::Error)?;
```
[the remark this answers is not in the document]
Comment (dmoka) [reply pallets/omnipool/src/router_execution.rs:161]: That might be indeed good to check.

On the other hand, we discussed that LRNA won't be supported in onchain routing.

image

[the remark this answers is not in the document]
Comment (mrq1911) [reply pallets/omnipool/src/router_execution.rs:161]: yeah, buying LRNA will always fail currently, therefore we cannot really store that route as we cannot validate it

**Review by mrq1911 — APPROVED** [review]

**Pull request merged**